lib/logstash/filters/pilar.rb

```diff
@@ -44,45 +44,56 @@ class Pilar < LogStash::Filters::Base
       # ['(\d+\.){3}\d+'] for IP addresses to be extracted before parsing begins.
       config :regexes, validate: :array, required: false, default: []
 
+      # This determines the maximum size of the single, double, and triple gram dictionaries respectively.
+      # Upon any of those hash maps reaching their maximum size, a LRU evicition policy is used to remove items.
+      # This controls the upper limit of the memory usage of this filter.
+      config :maximum_gram_dict_size, validate: :number, required: false, default: 10_000
+
       def register
         @linenumber = 1
         @regexes = regexes.map { |regex| Regexp.new(regex) }
-        @gramdict = GramDict.new
-        @preprocessor = Preprocessor.new(@gramdict, @logformat, @content_specifier, @regexes)
 
         # Check if dynamic_token_threshold is between 0 and 1
-        if @dynamic_token_threshold < 0.0 || @dynamic_token_threshold > 1.0
-          raise LogStash::ConfigurationError, 'dynamic_token_threshold must be between 0 and 1'
-        end
+        return unless @dynamic_token_threshold < 0.0 || @dynamic_token_threshold > 1.0
 
-        # populate gramdict with seed logs
-        return unless @seed_logs_path
+        raise LogStash::ConfigurationError, 'dynamic_token_threshold must be between 0 and 1'
+      end
 
-        ::File.open(@seed_logs_path, 'r') do |seed_logs|
-          seed_logs.each_line do |seed_log|
-            # TODO: Here, we are parsing every seed log file when we don't need to,
-            # might need to separate these steps out
-            @preprocessor.process_log_event(seed_log, false)
+      def filter(event)
+        # Initialize gramdict and preprocessor for this thread if not already done
+        unless Thread.current[:gramdict] && Thread.current[:preprocessor]
```

Comment:
thread local variable logic. Here, I create a `gramdict` and `preprocessor` per worker and run the seed log for each worker.